File: count-of-smaller-numbers-after-self/count-of-smaller-numbers-after-self-1.java
class Solution {

    private int[] index;
    private int[] temp;
    private int[] tempIndex;
    private int[] res;

    public List<Integer> countSmaller(int[] nums) {
        List<Integer> result = new ArrayList<>();

        this.index = new int[nums.length];
        this.temp = new int[nums.length];
        this.tempIndex = new int[nums.length];
        this.res = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            index[i] = i;
        }

        int l = 0, r = nums.length - 1;
        mergeSort(nums, l, r);

        for (int i = 0; i < nums.length; i++) {
            result.add(res[i]);
        }

        return result;
    }

    private void mergeSort(int[] a, int l, int r) {
        if (l >= r) {
            return;
        }

        int mid = (l + r) >> 1;
        mergeSort(a, l, mid);
        mergeSort(a, mid+1, r);
        merge(a, l, mid, r);
    }

    private void merge(int[] a, int l, int mid, int r) {
        int i = l, j = mid + 1, p = l;
        while (i <= mid && j <= r) {
            if (a[i] <= a[j]) {
                temp[p] = a[i];
                tempIndex[p] = index[i];
                res[index[i]] += (j - mid - 1);
                i++;
                p++;
            } else {
                temp[p] = a[j];
                tempIndex[p] = index[j];
                j++;
                p++;
            }
        }

        while (i <= mid) {
            temp[p] = a[i];
            tempIndex[p] = index[i];
            res[index[i]] += (j - mid - 1);
            i++;
            p++;
        }

        while (j <= r) {
            temp[p] = a[j];
            tempIndex[p] = index[j];
            j++;
            p++;
        }

        for (int k = l; k <= r; k++) {
            index[k] = tempIndex[k];
            a[k] = temp[k];
        }
    }
}

File: count-of-smaller-numbers-after-self/count-of-smaller-numbers-after-self-1.c
void merge(int *nums, int left, int mid, int right, int* result, int* index, int* temp, int* tempIndex) {
    int i = left, j = mid + 1, p = left;
    while (i <= mid && j <= right) {
        if (nums[i] <= nums[j]) {
            temp[p] = nums[i];
            tempIndex[p] = index[i];
            result[index[i]] += j - (mid + 1);
            i++;
            p++;
        } else {
            temp[p] = nums[j];
            tempIndex[p] = index[j];
            j++;
            p++;
        }
    }

    while (i <= mid) {
        temp[p] = nums[i];
        tempIndex[p] = index[i];
        result[index[i]] += right - mid;
        i++;
        p++;
    }

    while (j <= right) {
        temp[p] = nums[j];
        tempIndex[p] = index[j];
        j++;
        p++;
    }

    for (int k = left; k <= right; k++) {
        nums[k] = temp[k];
        index[k] = tempIndex[k];
    }
}

void mergeSort(int* nums, int left, int right, int* result, int* index, int* temp, int* tempIndex) {
    if (left == right) {
        return;
    }

    int mid = left + (right - left) / 2;
    mergeSort(nums, left, mid, result, index, temp, tempIndex);
    mergeSort(nums, mid + 1, right, result, index, temp, tempIndex);

    if (nums[mid] <= nums[mid+1]) {
        return;
    }

    merge(nums, left, mid, right, result, index, temp, tempIndex);
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* countSmaller(int* nums, int numsSize, int* returnSize){

    int* result = calloc(sizeof(int), numsSize);
    *returnSize = numsSize;

    if (nums == NULL || numsSize < 1) {
        return NULL;
    }

    int* index = malloc(sizeof(int) * numsSize);
    int* temp = malloc(sizeof(int) * numsSize);
    int* tempIndex = malloc(sizeof(int) * numsSize);

    for (int i = 0; i < numsSize; i++) {
        index[i] = i;
    }

    mergeSort(nums, 0, numsSize - 1, result, index, temp, tempIndex);

    free(index);
    free(temp);
    free(tempIndex);

    return result;
}